### ISSUE TITLE
Problem with masked dataset in detrend preprocessor

### ISSUE THREAD

**Issue by stefsmeets** [ESMValGroup/ESMValCore#768]
**Describe the bug**
When implementing a new recipe (https://github.com/ESMValGroup/ESMValTool/pull/1648) we came accross an problem when using a CMORized E-OBS dataset.

Specifically once we added the detrend preprocessor, an error was raised because of NaNs in the data. This might be a problem with the CMORizer as suggested here https://github.com/ESMValGroup/ESMValTool/pull/1648#discussion_r477202617

Note that the script otherwise works fine with another observation dataset that we were using (ERA5/native6).

**Please attach**
- [recipe_climwip.txt](https://github.com/ESMValGroup/ESMValTool/files/5142129/recipe_climwip.txt)
- [main_log_debug.txt](https://github.com/ESMValGroup/ESMValTool/files/5142118/main_log_debug.txt)

**Comment by bouweandela** [comment]
@jvegasbsc Would you be able to have a look at this?

**Comment by valeriupredoi** [comment]
indeed, `detrend` doesn't deal with masked objects, but since the process is just removing the line of least squares, one can just give the NaN a value that's outrageous and no being taken by any other valid datum eg 1e20

**Comment by sloosvel** [comment]
@pepcos had a similar issue. If adding back the mask after computing `detrended` is a good enough solution, it could be done with something along the lines of `da.ma.getmaskarray` and `da.ma.mask_where`.

**Comment by bouweandela** [comment]
I doubt whether that is good enough, because the masked values should not contribute to the trend. For example:
```python
>>> a = np.ma.array([100, 1, 2, 3, 100], mask=[True, False, False, False, True])
>>> a
masked_array(data=[--, 1, 2, 3, --],
             mask=[ True, False, False, False,  True],
       fill_value=999999)
>>> b = scipy.signal.detrend(a)
```
gives
```python
>>> b
array([ 59.2, -40. , -39.2, -38.4,  58.4])
```
while it should be
```
[--, 0, 0, 0, --]
```

**Comment by sloosvel** [comment]
Maybe something like:

`scipy.signal.detrend(np.argwhere(a.mask==False))`

and then rebuilding the output array could to the trick

**Comment by bouweandela** [comment]
I think you would need to take the distance between the points into account to compute the trend, just dropping some points will give you wrong results. Maybe compute the trend with https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.linregress.html and then subtract that from the input array?

**Comment by sloosvel** [comment]
Thanks! I get some differences with that approach though. Does anybody have the time to check whether they are reasonable because of using a different method?

**Comment by bouweandela** [comment]
Maybe someone from our @ESMValGroup/science-reviewers team?
